### PR TITLE
Stages rework

### PR DESCRIPTION
**Pull request by PiotrZierhoffer**
This is a follow-up to #31.

Travis does not allow access to secret variables if the PR comes from another repo. This is a problem I don't know how to handle in a super robust way at the moment, but I think it's ok for now.

### CODE REVIEW

**Comment by PiotrZierhoffer** [comment]
For reference: https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions

**Comment by PiotrZierhoffer** [comment]
Doh. Anaconda requires an alphabetical prefix in label names, my uploads were landing on the main channel. Nothing got hurt, but still, irritating.

**Comment by PiotrZierhoffer** [comment]
One problem I don't know yet how to solve is cleanup - we upload unconditionally, each and every packet. After some time we'll have tons of labels that are no longer useful.

I could use some kind of "cleanup stage" after other stages, but there is no such concept in Travis (or at least - I'm not aware of it). The only alternative I know of is to set all packages as allowed to fail, so that all stages pass, and have the last stage that removes obsolete packages. This will make build results useless.

**Comment by mithro** [comment]
We should just have a bot which runs separately somewhere and removes any labels which are over 30 days old? (Maybe a github action?)

**Comment by PiotrZierhoffer** [comment]
Yeah, it's possible to list labels and packages within labels with 

```
anaconda -t $TOKEN label --list -o litex-hub
anaconda -t $TOKEN label -o litex-hub --show label-name
```

You can show package info with `anaconda show $PACKAGE_NAME`, but it's a pain to parse. 

I'll think about this.

**Comment by PiotrZierhoffer** [comment]
@mithro we're close to being green here. So now it's about stage names. In this particular case we have "standalone packages" and "yosys-based packages". Is this the kind of naming you'd want?

File: .travis.yml [hunk 1]
```diff
@@ -6,162 +6,157 @@ addons:
...
 
 stages:
-  - "EDA Tools - Misc"
-  - "EDA Tools - Bitstream"
-  - "EDA Tools - Simulation"
-  - "EDA Tools - Synthesis"
-  - "EDA Tools - PnR"
-  - "EDA Tools - Formal"
-  - "EDA Tools - Xilinx Vivado metapackages"
+  - "Stage one"
```
Comment (mithro) [inline .travis.yml:9]:
```suggestion
  - "No dependencies" # Doesn't depend on anything apart from upstream packages.
  - "Has first level dependencies" # Depends on a package in the "no dependencies" group
  - "Has second level dependencies" # Depends on a package in the "Has first level dependencies" group
  - ....
```

**Comment by mithro** [comment]
@PiotrZierhoffer I added a potential suggestion, open to other ideas!

**Comment by mithro** [comment]
@PiotrZierhoffer - I'm surprised at how simple this turned out to be!

**Comment by mithro** [comment]
@PiotrZierhoffer - There should be a final build stage which "upgrades" the packages from the travis channel to the master one, right?

**Comment by PiotrZierhoffer** [comment]
@mithro ack for stages names, will implement it shortly
Regarding master - the packages are pushed to the main label just after they're are sent to the temporary label - but only if the build is started on master branch. So the flow is more or less the same as it used to be

**Comment by mithro** [comment]
@PiotrZierhoffer - We should only push the packages to main after everything is built?

**Comment by PiotrZierhoffer** [comment]
I can't really see the reason for that. If the package is built, then up to our best knowledge it's OK. If subsequent packages fail, then it's either not related, or the base package moved forward and is now incompatible with the one that fails. Again, this is a problem of the failing package, not the one that was successfully built.

These packages are mostly unrelated, so I don't think we should introduce this kind of relationship between them.

**Comment by mithro** [comment]
I don't think we want to push a new package which breaks a dependent package?

**Comment by PiotrZierhoffer** [comment]
ok, i'll give it a shot.

**Comment by mithro** [comment]
@PiotrZierhoffer -- It is also the first step to moving to a system where we have;

 Step 1 - building + uploading
 Step 2 - download and test it works --> If it works, mark as good.

**Comment by PiotrZierhoffer** [comment]
I took the travis cron job approach. I have not yet defined it, because we'd want to have this merged first.

@mithro please take a look

**Comment by mithro** [comment]
FYI - There is a weird error in the build output @ https://travis-ci.com/github/litex-hub/litex-conda-eda/jobs/359921093 -- I think it is unrelated to this change however....
```
 pnr/arachne: -  Downloading..
 -------------------------------------------------------------------
Still running (1 of 20): conda build --source pnr/arachne
The command conda build --source pnr/arachne exited with 1.
Log:
No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
WARNING:conda_build.metadata:No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.11
Cloning into bare repository '/tmp/really-long-path/conda/conda-bld/git_cache/github.com/mithro/arachne-pnr.git'...
Cloning into '/tmp/really-long-path/conda/conda-bld/arachne-pnr_1594404073933/work'...
done.
Already on 'master'
Your branch is up to date with 'origin/master'.
commit f4a52e2ad28c01960979548954d494eb23a19621
Author: Tim 'mithro' Ansell <me@mith.ro>
Date:   Tue Aug 14 19:30:47 2018 -0700
    Fix tri.v for updated icestorm.
    
    Fixes #120.
    
    Signed-off-by: Tim 'mithro' Ansell <me@mith.ro>
v0.0-310-gf4a52e2
On branch master
Your branch is up to date with 'origin/master'.
nothing to commit, working tree clean
checkout: 'master'
==> git log -n1 <==
==> git describe --tags --dirty <==
==> git status <==
Adding in variants from internal_defaults
INFO:conda_build.variants:Adding in variants from internal_defaults
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... done
Attempting to finalize metadata for arachne-pnr
INFO:conda_build.metadata:Attempting to finalize metadata for arachne-pnr
INFO conda_build.metadata:finalize_outputs_pass(700): Attempting to finalize metadata for arachne-pnr
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... done
fatal: destination path '/tmp/really-long-path/conda/conda-bld/arachne-pnr_1594404073933/work' already exists and is not an empty directory.
Traceback (most recent call last):
  File "/tmp/really-long-path/conda/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 469, in main
    execute(sys.argv[1:])
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in execute
    outputs = [action(recipe, config) for recipe in args.recipe]
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in <listcomp>
    outputs = [action(recipe, config) for recipe in args.recipe]
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 385, in source_action
    source.provide(metadata)
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/source.py", line 655, in provide
    verbose=metadata.config.verbose)
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/source.py", line 328, in git_source
    git_depth=git_depth, is_top_level=True, verbose=verbose)
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/source.py", line 249, in git_mirror_checkout_recursive
    check_call_env([git, 'clone', git_mirror_dir, git_checkout_dir], stdout=stdout, stderr=stderr)
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/utils.py", line 398, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/tmp/really-long-path/conda/lib/python3.7/site-packages/conda_build/utils.py", line 389, in _func_defaulting_env_to_os_environ
    subprocess.check_call(_args, **kwargs)
  File "/tmp/really-long-path/conda/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/git', 'clone', '/tmp/really-long-path/conda/conda-bld/git_cache/github.com/mithro/arachne-pnr.git', '/tmp/really-long-path/conda/conda-bld/arachne-pnr_1594404073933/work']' returned non-zero exit status 128.
./.travis/install.sh: line 1:  8154 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
```

File: .travis.yml [hunk 2]
```diff
@@ -6,162 +6,166 @@ addons:
...
-  - "EDA Tools - Synthesis"
-  - "EDA Tools - PnR"
-  - "EDA Tools - Formal"
-  - "EDA Tools - Xilinx Vivado metapackages"
+  - name: "No dependencies" # Doesn't depend on anything apart from upstream packages
+    if: type != cron
+  - name: "Has first level dependencies" # Depends on a package in the "no dependencies" group
+    if: type != cron
+  - name: "Main label upload" # Puts Anaconda packages to the main label
+    if: type = push AND \
```
Comment (mithro) [inline .travis.yml:14]: Why not keep this one line?

File: .travis.yml [hunk 3]
```diff
@@ -172,16 +176,30 @@ jobs:
     after_success:
       - echo "Xilinx Vivado metapackages built for versions:" && cat $VERSIONS
 
+ # Move packages from the current label to main
```
Comment (mithro) [inline .travis.yml:179]: I don't see this running on https://travis-ci.com/github/litex-hub/litex-conda-eda/builds/175201099 ?

File: .travis/cleanup-anaconda.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (mithro) [inline .travis/cleanup-anaconda.sh:1]: Would this be better as a Python script (rather than bash?).

File: .travis/master-package.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (mithro) [inline .travis/master-package.sh:1]: Would this be better as a Python script (rather than bash?).

File: .travis/after_success.sh [hunk 6]
```diff
@@ -11,14 +11,14 @@ if [[ $UPLOAD == "no-upload" ]]; then
...
-    	end_section "package.upload"
+    $SPACER
+    # Travis will not expose the ANACONDA_TOKEN var for pull requests coming from other forks than the original one
+    if [ x$ANACONDA_TOKEN != x ]; then
+        start_section "package.upload" "${GREEN}Package uploading...${NC}"
+        anaconda -t $ANACONDA_TOKEN upload --no-progress --user $ANACONDA_USER --label travis-$TRAVIS_BUILD_ID $CONDA_OUT
+        end_section "package.upload"
     else
-        echo "Conditions unmet, will not upload"
+        echo "Cannot upload packages from branches not uploaded to the original repository"
```
Comment (mithro) [inline .travis/after_success.sh:21]: I think we want to say something like "Please define ANACONDA_TOKEN in your travis build settings (url)"? or something similar?

File: .travis.yml [hunk 7]
```diff
@@ -6,162 +6,166 @@ addons:
...
-  - "EDA Tools - PnR"
-  - "EDA Tools - Formal"
-  - "EDA Tools - Xilinx Vivado metapackages"
+  - name: "No dependencies" # Doesn't depend on anything apart from upstream packages
+    if: type != cron
+  - name: "Has first level dependencies" # Depends on a package in the "no dependencies" group
+    if: type != cron
+  - name: "Main label upload" # Puts Anaconda packages to the main label
+    if: type = push AND \
+        branch = master
```
Comment (mithro) [inline .travis.yml:15]: Maybe we should push to a label named after the branch?

File: .travis.yml (hunk 3 above)
Comment (PiotrZierhoffer) [reply .travis.yml:179]: only on master

File: .travis/after_success.sh (hunk 6 above)
Comment (PiotrZierhoffer) [reply .travis/after_success.sh:21]: This is not really true. We have this defined, but it's a secret variable. Travis will not expose it to any branch from another fork. It's a security feature. There's nothing people could do. It's a serious problem from the development perspective.

File: .travis.yml (hunk 2 above)
Comment (PiotrZierhoffer) [reply .travis.yml:14]: no particular reason, fixed

File: .travis.yml (hunk 7 above)
Comment (PiotrZierhoffer) [reply .travis.yml:15]: travis-{branch}-{run}? Sure, makes sense

**Comment by PiotrZierhoffer** [comment]
I push commits, yet for some reason the PR remains unchanged. I believe it's a github problem

File: .travis/after_success.sh (hunk 6 above)
Comment (mithro) [reply .travis/after_success.sh:21]: On a PR the token won't be available, but when people push to their own fork the tool should be uploading to that user's own anaconda repo which they can get an `ANACONDA_TOKEN` for.

Setting up their own anaconda repo and `ANACODNA_TOKEN` should be explained somewhere....

Comment (PiotrZierhoffer) [reply .travis/after_success.sh:21]: understood

**Pull request merged**